### PR TITLE
DOCSP-34351 PostGres Migration Prep Docs

### PR DESCRIPTION
**Pull request by ianf-mongodb**
## DESCRIPTION

PostGreSQL migration prep task page. Based off the description of https://jira.mongodb.org/browse/DOCSP-34351 and 
https://debezium.io/documentation/reference/stable/connectors/postgresql.html.

## STAGING

https://docs-mongodbcom-staging.corp.mongodb.com/docs-relational-migrator/docsworker-xlarge/DOCSP-34351/prerequisites/postgres/

## JIRA

https://jira.mongodb.org/browse/DOCSP-34351

## BUILD LOG

https://workerpool-boxgs.mongodbstitch.com/pages/job.html?collName=queue&jobId=65cce858cbcd269fa527d2a3

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Is this free of spelling errors?
- [x] Is this free of grammatical errors?
- [x] Is this free of staging / rendering issues?
- [x] Are all the links working?

## External Review Requirements

[What's expected of an external reviewer?](https://wiki.corp.mongodb.com/display/DE/Reviewing+Guidelines+for+the+MongoDB+Server+Documentation)

### CODE REVIEW

File: source/prerequisites/postgres.txt [hunk 1]
```diff
@@ -10,4 +10,78 @@ Postgres Migration Prerequisites
...
+
+         wal_level = logical 
+
+   .. step:: Create a SQL Replication Login
+
+      The PostgreSQL account requires ``REPLICATION`` and ``LOGIN`` 
+      database permissions as well as ``SELECT`` and ``USAGE`` 
+      permissions on each source table:
+
+      .. code-block:: sql
```
Comment (ianf-mongodb) [inline source/prerequisites/postgres.txt:68]: For Tech Review:

Should we add example grant statements for the permissions mentioned above?

File: source/prerequisites/postgres.txt [hunk 2]
```diff
@@ -1,13 +1,87 @@
...
\ No newline at end of file
+About this Task
+---------------
+
+Connecting Relational Migrator to PostgreSQL requires
+`logical decoding 
+<https://www.postgresql.org/docs/current/logicaldecoding-explanation.html>`__ 
+enabled on your PostGreSQL instance. When you set up a new connection, 
+Relational Migrator automatically generates the required SQL code to 
+configure PostgreSQL. This document explains the purpose of those scripts.
```
Comment (jocelyn-mendez1) [inline source/prerequisites/postgres.txt:21]: [Suggestion]

I feel like this makes it seem like the procedure is about explaining the scripts but if I understand correctly it's configuring PostgreSQL and along the way the purpose of those scripts is explained? If so, maybe we can reword this to explain that by doing this procedure you also understand the scripts that are produced.

**Review by jocelyn-mendez1 — CHANGES_REQUESTED** [review]
Thanks for these updates @ianf-mongodb! I've left some suggestions so please feel free to let me know if I can expand on anything!

File: source/prerequisites/postgres.txt [hunk 3]
```diff
@@ -1,13 +1,87 @@
...
 
 .. contents:: On this page
    :local:
    :backlinks: none
    :depth: 1
    :class: singlecol
 
-Postgres.
\ No newline at end of file
+About this Task
```
[the remark this answers is not in the document]
Comment (ianf-mongodb) [reply source/prerequisites/postgres.txt]: Removed.

File: source/prerequisites/sql-server.txt [hunk 4]
```diff
@@ -69,7 +69,7 @@ job configurations.
           unchanged, and does not add any performance overhead. Enabling
           CDC alone does not result in changes being captured.
 
-            .. include:: /includes/fact-enable-cdc-database.rst
+         .. include:: /includes/fact-enable-cdc-database.rst
```
Comment (ianf-mongodb) [inline source/prerequisites/sql-server.txt:72]: **Ignore** Fixing build error.

**Review by ianf-mongodb — COMMENTED** [review]
Thanks for the review @jocelyn-mendez1 this is ready for another look!

File: source/prerequisites/postgres.txt [hunk 5]
```diff
@@ -1,13 +1,84 @@
 .. _rm-prereq-postgres:
 
-================================
-Postgres Migration Prerequisites
-================================
+==================================
```
Comment (jocelyn-mendez1) [inline source/prerequisites/postgres.txt:3]: [Info Typing]

Task page titles should be verb phrases.

File: source/prerequisites/postgres.txt [hunk 6]
```diff
@@ -1,13 +1,84 @@
...
-Postgres.
\ No newline at end of file
+Connecting Relational Migrator to PostgreSQL requires
+`logical decoding 
+<https://www.postgresql.org/docs/current/logicaldecoding-explanation.html>`__ 
+to be enabled on your PostgreSQL instance. When you set up a new connection, 
+Relational Migrator automatically generates the required SQL code to 
+configure PostgreSQL. This document explains the purpose of those scripts.
+
+These instructions configure PostgreSQL **for both** Snapshot and
```
Comment (jocelyn-mendez1) [inline source/prerequisites/postgres.txt:20]: [Italics]

```suggestion
These instructions configure PostgreSQL *for both* Snapshot and 
```

File: source/prerequisites/postgres.txt [hunk 7]
```diff
@@ -1,13 +1,84 @@
...
+`logical decoding 
+<https://www.postgresql.org/docs/current/logicaldecoding-explanation.html>`__ 
+to be enabled on your PostgreSQL instance. When you set up a new connection, 
+Relational Migrator automatically generates the required SQL code to 
+configure PostgreSQL. This document explains the purpose of those scripts.
+
+These instructions configure PostgreSQL **for both** Snapshot and 
+Continuous sync jobs.
+
+- Snapshot sync jobs migrate all the data and then stops.
```
Comment (jocelyn-mendez1) [inline source/prerequisites/postgres.txt:23]:
```suggestion
- Snapshot sync jobs migrate all the data and then stop.
```

File: source/prerequisites/postgres.txt [hunk 8]
```diff
@@ -1,13 +1,84 @@
...
+
+      The PostgreSQL account requires ``REPLICATION`` and ``LOGIN`` 
+      database permissions as well as ``SELECT`` and ``USAGE`` 
+      permissions on each source table:
+
+      .. code-block:: sql
+
+         CREATE ROLE <name> REPLICATION LOGIN;
+
+   .. step:: Create a Publication
```
Comment (jocelyn-mendez1) [inline source/prerequisites/postgres.txt:69]:
```suggestion
   .. step:: Create a publication
```

File: source/prerequisites/postgres.txt [hunk 9]
```diff
@@ -1,13 +1,84 @@
...
+<https://www.postgresql.org/docs/current/logicaldecoding-explanation.html>`__ 
+to be enabled on your PostgreSQL instance. When you set up a new connection, 
+Relational Migrator automatically generates the required SQL code to 
+configure PostgreSQL. This document explains the purpose of those scripts.
+
+These instructions configure PostgreSQL **for both** Snapshot and 
+Continuous sync jobs.
+
+- Snapshot sync jobs migrate all the data and then stops.
+- Continuous sync job run a snapshot and then enter a CDC stage to
```
Comment (jocelyn-mendez1) [inline source/prerequisites/postgres.txt:24]:
```suggestion
- Continuous sync jobs run a snapshot and then enter a CDC stage to 
```

**Review by jocelyn-mendez1 — CHANGES_REQUESTED** [review]
Hi @ianf-mongodb thanks for the updates! I just left some smaller suggestions so let me know if you have any questions!

File: source/prerequisites/postgres.txt (hunk 5 above)
Comment (ianf-mongodb) [reply source/prerequisites/postgres.txt:3]: Will ask Tom about a preferred name for these pages that is more verb driven and update accordingly.

**Review by ianf-mongodb — COMMENTED** [review]
@jocelyn-mendez1 Thanks for your feedback this is ready for another look %1 pending comment which I will be checking before merging this feature branch.

**Comment by ianf-mongodb** [comment]
@tomhollander could you take a look at these when you get a moment please.

File: source/prerequisites/postgres.txt (hunk 5 above)
Comment (jocelyn-mendez1) [reply source/prerequisites/postgres.txt:3]: Sounds good, thanks Ian!

**Review by jocelyn-mendez1 — APPROVED** [review]
Thanks for the updates @ianf-mongodb, this LGTM!

**Review by tomhollander — CHANGES_REQUESTED** [review]
The wal_level and publications are only needed for continuous jobs. For snapshot jobs there are no pre-reqs.

The CDC steps are incomplete. Please see the attached generated file for an example of all the steps that may be needed, specifically adding the tables to the replication groups and configuring REPLICA IDENTITY. 
[awards_POSTGRESQL_config.sql.txt](https://github.com/mongodb/docs-relational-migrator/files/14214569/awards_POSTGRESQL_config.sql.txt)

File: source/prerequisites/postgres.txt [hunk 10]
```diff
@@ -1,13 +1,147 @@
...
+
+      For snapshot jobs against PostGreSQL, the service account requires
+      schema ``USAGE`` and table ``SELECT`` permissions. Replace 
+      ``POSTGRES`` with the name of your service account user:
+
+      .. code-block:: sql
+         :copyable: true
+
+         GRANT USAGE ON SCHEMA SCHEMA_NAME TO POSTGRES;
+         GRANT SELECT ON TABLE SCHEMA_NAME.TABLE_NAME TO POSTGRES;
```
Comment (ianf-mongodb) [inline source/prerequisites/postgres.txt:52]: I confirmed with Poonam that schema usage and select is required even for snapshot: https://github.com/mongodb-ets/migrator/blob/main/application/src/main/resources/script/postgresql.mustache#L35-L42.

**Review by ianf-mongodb — COMMENTED** [review]
Hi @tomhollander I have updated the procedure based on the script you provider and the postGreSQL mustache files. Please LMKWYT Thanks!

File: source/prerequisites/postgres.txt [hunk 11]
```diff
@@ -1,13 +1,147 @@
...
 
 .. contents:: On this page
    :local:
    :backlinks: none
    :depth: 1
    :class: singlecol
 
-Postgres.
\ No newline at end of file
+To run sync jobs from an PostGreSQL source database, the database may
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:13]: Product name is "PostgreSQL" - no capital G. Needs to be fixed throughout.

File: source/prerequisites/postgres.txt [hunk 12]
```diff
@@ -1,13 +1,147 @@
...
+configuration PostGreSQL for both types of sync jobs:
+
+.. include:: /includes/fact-short-sync-job-desc.rst
+
+About this Task
+---------------
+
+Relational Migrator uses the PostgreSQL ``pgoutput`` plugin for logical 
+decoding. This plug-in is included in PostgreSQL versions 10+, no 
+additional installation or configuration is required.
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:28]: I don't think this text helps anyone. RM doesn't support anything below v10 anyway, and we don't explain why this is relevant. I'd just cut this out.

File: source/prerequisites/postgres.txt (hunk 10 above)
Comment (tomhollander) [reply source/prerequisites/postgres.txt:52]: OK. One thing to note is that Postgres is really funny about names that aren't all lowercase. While I know it's an example, can we change the sample code to:
```
GRANT USAGE ON SCHEMA <schema_name> TO <database_user_name>;
GRANT SELECT ON TABLE <schema_name>.<table_name> TO <database_user_name>;
```
This also makes it more obvious what needs to be changed.

File: source/prerequisites/postgres.txt [hunk 13]
```diff
@@ -1,13 +1,147 @@
...
+         account, and create a publication.
+
+         .. step:: Edit the configuration file 
+
+            Logical replication may not be enabled by default. To enable logical 
+            replication, change the `wal_level <https://postgresqlco.nf/doc/en/param/wal_level>`__ 
+            configuration in the 
+            `postgresql.conf configuration file <https://www.postgresql.org/docs/current/config-setting.html>`__. 
+            You *must restart the database instance* after changing the configuration file. 
+
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:71]: Can we add a quick note on how to do this in AWS RDS? The config file isn't exposed so you need to edit the RDS instance parameters. See 
https://dba.stackexchange.com/a/243576

File: source/prerequisites/postgres.txt [hunk 14]
```diff
@@ -1,13 +1,147 @@
...
+                  CREATE ROLE REPLICATION_GROUP REPLICATION LOGIN;
+
+            #. Grant the table ``SELECT`` and schema ``USAGE``
+               permissions to the ``REPLICATION_GROUP`` role. Each table in
+               the migration requires a ``GRANT SELECT`` statement:
+
+               .. code-block:: sql
+                  :copyable: true
+
+                  GRANT USAGE ON SCHEMA SCHEMA_NAME TO REPLICATION_GROUP;
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:92]: As above, can we use `<schema_name>` etc so you can tell what's a keyword and what needs to be replaced.

File: source/prerequisites/postgres.txt [hunk 15]
```diff
@@ -1,13 +1,147 @@
...
+            #. Grant the REPLICATION_GROUP role to the service account
+
+               Replace ``ORIGINAL_OWNER`` with the owner of the 
+               participating tables. Replace ``POSTGRES`` with the name of 
+               your service account user:
+
+               .. code-block:: sql
+                  :copyable: true
+
+                  GRANT REPLICATION_GROUP TO ORIGINAL_OWNER;
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:105]: As above

File: source/prerequisites/postgres.txt [hunk 16]
```diff
@@ -1,13 +1,147 @@
...
+         .. step:: Grant ownership of each table to the REPLICATION_GROUP role
+            
+            Each table in the migration requires a ``ALTER TABLE`` 
+            statement:
+
+            .. code-block:: sql
+               :copyable: true
+
+               ALTER TABLE SCHEMA.TABLE OWNER TO REPLICATION_GROUP;
+               -- ADDITIONAL ALTER TABLE STATEMENTS...
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:117]: As above

File: source/prerequisites/postgres.txt [hunk 17]
```diff
@@ -1,13 +1,147 @@
...
+
+            Create a `publication <https://www.postgresql.org/docs/current/logical-replication-publication.html>`__ 
+            each ``SCHEMA.TABLE`` in the migration must be specified 
+            in the ``FOR`` statement separated by commas:
+
+            .. code-block:: sql
+               :copyable: true
+
+               CREATE PUBLICATION "MIGRATOR_NAME_PUBLICATION" 
+               FOR TABLE "SCHEMA"."TABLE","SCHEMA"."TABLE2";
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:129]: As above

File: source/prerequisites/postgres.txt [hunk 18]
```diff
@@ -1,13 +1,147 @@
...
+
+         .. step:: Set replica identity to full
+
+            Each table in the migration requires a ``ALTER TABLE`` 
+            statement:
+
+            .. code-block:: sql
+               :copyable: true
+
+               ALTER TABLE SCHEMA.TABLE REPLICA IDENTITY FULL;
```
Comment (tomhollander) [inline source/prerequisites/postgres.txt:139]: As above

**Review by tomhollander — CHANGES_REQUESTED** [review]
Good overall, suggested a few things to tidy up.

File: source/prerequisites/postgres.txt (hunk 12 above)
Comment (ianf-mongodb) [reply source/prerequisites/postgres.txt:28]: Done.

File: source/prerequisites/postgres.txt (hunk 10 above)
Comment (ianf-mongodb) [reply source/prerequisites/postgres.txt:52]: Updated.

File: source/prerequisites/postgres.txt (hunk 13 above)
Comment (ianf-mongodb) [reply source/prerequisites/postgres.txt:71]: I added a link and some context to how to do this in AWS RDS to this step.

File: source/prerequisites/postgres.txt (hunk 14 above)
Comment (ianf-mongodb) [reply source/prerequisites/postgres.txt:92]: Updated throughout

**Review by ianf-mongodb — COMMENTED** [review]
Thanks @tomhollander your feedback has been implemented. Ready for another look.

**Review by tomhollander — APPROVED** [review]
Thanks for the updates!
